node/cn/peer_set.go

```diff
@@ -0,0 +1,436 @@
+type peerSet struct {
+	peers   map[string]Peer
+	cnpeers map[common.Address]Peer
+	pnpeers map[common.Address]Peer
+	enpeers map[common.Address]Peer
+	lock    sync.RWMutex
+	closed  bool
+
+	validator map[p2p.ConnType]p2p.PeerTypeValidator
+}
+
+// newPeerSet creates a new peer set to track the active participants.
+func newPeerSet() *peerSet {
+	peerSet := &peerSet{
+		peers:     make(map[string]Peer),
+		cnpeers:   make(map[common.Address]Peer),
+		pnpeers:   make(map[common.Address]Peer),
+		enpeers:   make(map[common.Address]Peer),
+		validator: make(map[p2p.ConnType]p2p.PeerTypeValidator),
+	}
+
+	peerSet.validator[node.CONSENSUSNODE] = ByPassValidator{}
+	peerSet.validator[node.PROXYNODE] = ByPassValidator{}
+	peerSet.validator[node.ENDPOINTNODE] = ByPassValidator{}
+
+	return peerSet
+}
+
+// Register injects a new peer into the working set, or returns an error if the
+// peer is already known.
+func (ps *peerSet) Register(p Peer) error {
+	ps.lock.Lock()
+	defer ps.lock.Unlock()
+
+	if ps.closed {
+		return errClosed
+	}
+	if _, ok := ps.peers[p.GetID()]; ok {
+		return errAlreadyRegistered
+	}
+
+	var peersByNodeType map[common.Address]Peer
+	var peerTypeValidator p2p.PeerTypeValidator
+
+	switch p.ConnType() {
+	case node.CONSENSUSNODE:
+		peersByNodeType = ps.cnpeers
+		peerTypeValidator = ps.validator[node.CONSENSUSNODE]
+	case node.PROXYNODE:
+		peersByNodeType = ps.pnpeers
+		peerTypeValidator = ps.validator[node.PROXYNODE]
+	case node.ENDPOINTNODE:
+		peersByNodeType = ps.enpeers
+		peerTypeValidator = ps.validator[node.ENDPOINTNODE]
+	default:
+		return fmt.Errorf("undefined peer type entered, p.ConnType(): %v", p.ConnType())
+	}
+
+	if _, ok := peersByNodeType[p.GetAddr()]; ok {
+		return errAlreadyRegistered
+	}
+
+	if err := peerTypeValidator.ValidatePeerType(p.GetAddr()); err != nil {
+		return fmt.Errorf("fail to validate peer type: %s", err)
+	}
+
+	peersByNodeType[p.GetAddr()] = p // add peer to its node type peer map.
+	ps.peers[p.GetID()] = p          // add peer to entire peer map.
+
+	cnPeerCountGauge.Update(int64(len(ps.cnpeers)))
+	pnPeerCountGauge.Update(int64(len(ps.pnpeers)))
+	enPeerCountGauge.Update(int64(len(ps.enpeers)))
+	go p.Broadcast()
+
+	return nil
+}
+
+// Unregister removes a remote peer from the active set, disabling any further
+// actions to/from that particular entity.
+func (ps *peerSet) Unregister(id string) error {
+	ps.lock.Lock()
+	defer ps.lock.Unlock()
+
+	p, ok := ps.peers[id]
+	if !ok {
+		return errNotRegistered
+	}
+	if p.ConnType() == node.CONSENSUSNODE {
+		delete(ps.cnpeers, p.GetAddr())
+	} else if p.ConnType() == node.PROXYNODE {
+		delete(ps.pnpeers, p.GetAddr())
+	} else if p.ConnType() == node.ENDPOINTNODE {
+		delete(ps.enpeers, p.GetAddr())
+	}
+	delete(ps.peers, id)
+	p.Close()
+
+	cnPeerCountGauge.Update(int64(len(ps.cnpeers)))
+	pnPeerCountGauge.Update(int64(len(ps.pnpeers)))
+	enPeerCountGauge.Update(int64(len(ps.enpeers)))
+	return nil
+}
+
+// istanbul BFT
```

Comment:
(duplicated) Do I know the meaning of this comment?

Reply:
This one can be removed.